daemon/src/close_cfds.rs

```diff
@@ -0,0 +1,417 @@
+            taker_leverage,
+            settlement_interval,
+            quantity_usd,
+            counterparty_network_identity,
+            role,
+            opening_fee,
+            initial_funding_rate,
+            initial_tx_fee_rate,
+        }: db::Cfd = cfd;
+
+        let n_contracts = quantity_usd
+            .try_into_u64()
+            .expect("number of contracts to fit into a u64");
+        let n_contracts = Contracts::new(n_contracts);
+
+        let (long_leverage, short_leverage) =
+            long_and_short_leverage(taker_leverage, role, position);
+
+        let initial_funding_fee = FundingFee::calculate(
+            initial_price,
+            quantity_usd,
+            long_leverage,
+            short_leverage,
+            initial_funding_rate,
+            SETTLEMENT_INTERVAL.whole_hours(),
+        )
+        .expect("values from db to be sane");
+
+        let fee_account = FeeAccount::new(position, role)
+            .add_opening_fee(opening_fee)
+            .add_funding_fee(initial_funding_fee);
+
+        Self {
+            id,
+            position,
+            initial_price,
+            taker_leverage,
+            n_contracts,
+            counterparty_network_identity,
+            role,
+            opening_fee,
+            initial_funding_rate,
+            initial_tx_fee_rate,
+            fee_account,
+            own_script_pubkey: None,
+            lock: None,
+            commit: None,
+            collaborative_settlement: None,
+            non_collaborative_settlement: None,
+            refund: None,
+            close_status: CloseStatus::NotClosed,
+            version: 0,
+        }
+    }
+
+    fn is_closed(&self) -> bool {
+        self.close_status != CloseStatus::NotClosed
+    }
+}
+
+impl CfdAggregate for Cfd {
+    type CtorArgs = ();
+
+    fn new(args: Self::CtorArgs, cfd: db::Cfd) -> Self {
+        Cfd::new(cfd)
+    }
+
+    fn apply(mut self, event: model::CfdEvent) -> Self {
+        use model::EventKind::*;
+        match event.event {
+            ContractSetupStarted => {}
+            ContractSetupCompleted { dlc } => {
+                let script_pubkey = dlc.lock.1.script_pubkey();
+                let OutPoint { txid, vout } = dlc
+                    .lock
+                    .0
+                    .outpoint(&script_pubkey)
+                    .expect("DLC to be in lock TX");
+
+                let txid = Txid::new(txid);
+                let dlc_vout = Vout::new(vout);
+
+                self.lock = Some(Lock { txid, dlc_vout });
+
+                self.own_script_pubkey = Some(dlc.script_pubkey_for(self.role));
+            }
+            ContractSetupFailed => {}
+            OfferRejected => {}
+            RolloverStarted => {}
+            RolloverAccepted => {}
+            RolloverRejected => {}
+            RolloverCompleted { dlc, funding_fee } => {
+                self.fee_account.add_funding_fee(funding_fee);
+            }
```

Comment:
Shouldn't we set the commit transaction here?

Reply:
Good catch!

Reply:
Actually, no. Rolling over doesn't publish a commit transaction. The `ClosedCfd` struct (and the corresponding db model) only records transactions that are final.